File: sdk/samples/all in on code/Visual Studio 2008/CSSL3DataBinding/DAL/DAL.cs
/****************************** Module Header ******************************\
* Module Name:  DAL.cs
* Project:      CSSL3DataBinding
* Copyright (c) Microsoft Corporation.
* 
* This module contains Data Access Layer code and custom validation class. 
* For the simplicity, it gets in-momory data.
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
* All other rights reserved.
* 
* History:
* * 11/26/2009 04:00 PM Allen Chen Created
\***************************************************************************/

using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using System.ComponentModel;
using System.Collections.ObjectModel;
using System.Collections;
using System.ComponentModel.DataAnnotations;

namespace CSSL3DataBinding.DAL
{
    public static class CustomerGenerator
    {  
        private static ObservableCollection<Customer> _customercollection=new ObservableCollection<Customer>() {
            new Customer(){ ID=1, Name="Allen Chen"},
            new Customer(){ ID=2, Name="Mick Sun"},
            new Customer(){ ID=3, Name="Jialiang Ge"}};
        private static Customer _customer=new Customer() { ID = 1, Name = "Allen Chen" };
      
        public static object GetSingleCustomer() 
        {
            return _customer;        
        }

        public static IEnumerable GetCustomersList() 
        {
            return _customercollection;
        }

    }

    public class Customer : INotifyPropertyChanged
    {
        public int _id;
        // Use custom validation class for this property
        [CustomValidation(typeof(MyValidation),"Validate")]
        public int ID
        {            
            get
            {
                return _id;
            }
            set
            {
               
                Validator.ValidateProperty(value,
                       new ValidationContext(this, null, null) { MemberName = "ID" });
                _id = value;
                if (PropertyChanged != null)
                {
                    PropertyChanged(this, new PropertyChangedEventArgs("ID"));
                }
            }
        }
        private string _name;
        // Use DataAnnotation to restrict the length of Name property
        [StringLength(30, MinimumLength = 1, ErrorMessage =
           "Name must be between 1 and 30 characters long.")]
        public string Name
        {
            get
            {
                return _name;
            }
            set
            {

                Validator.ValidateProperty(value,
                       new ValidationContext(this, null, null) { MemberName = "Name" });
                _name = value;
                if (PropertyChanged != null)
                {
                    PropertyChanged(this, new PropertyChangedEventArgs("Name"));
                }
            }
        }

        #region INotifyPropertyChanged Members

        public event PropertyChangedEventHandler PropertyChanged;

        #endregion
    }

    public static class MyValidation
    {
        public static ValidationResult Validate(object property,
          ValidationContext context)
        {
            // You can get Property name from context.MemberName. It's not used here for simplicity.
            int customerid;
            // For the simplicity, the change is invalid if new ID equals 11.
            if (Int32.TryParse(property.ToString(),out customerid) && customerid == 11)
            {
                ValidationResult validationResult = new ValidationResult("Custom Validation Failed. ID cannot be 11");
                return validationResult;
            }
            else 
            {
                return ValidationResult.Success;
            }
        }
    }

}

File: sdk/samples/all in on code/Visual Studio 2008/VBSL3DataBinding/DAL/DAL.vb
'***************************** Module Header ******************************\
'* Module Name:	DAL.vb
'* Project:		VBSL3DataBinding
'* Copyright (c) Microsoft Corporation.
'* 
'* This module contains Data Access Layer code and custom validation class. 
'* For the simplicity, it gets in-momory data.
'* 
'* This source is subject to the Microsoft Public License.
'* See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
'* All other rights reserved.
'* 
'* History:
'* * 11/27/2009 11:00 AM Allen Chen Created
'\**************************************************************************
Imports System
Imports System.Net
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Documents
Imports System.Windows.Ink
Imports System.Windows.Input
Imports System.Windows.Media
Imports System.Windows.Media.Animation
Imports System.Windows.Shapes
Imports System.ComponentModel
Imports System.Collections.ObjectModel
Imports System.Collections
Imports System.ComponentModel.DataAnnotations

Namespace DAL
    Public Class CustomerGenerator
        Shared Sub New()
            _customer.ID = 1
            _customer.Name = "Allen Chen"
            _customercollection = New ObservableCollection(Of Customer)()
            Dim c1 As New Customer
            c1.ID = 1
            c1.Name = "Allen Chen"
            Dim c2 As New Customer
            c2.ID = 2
            c2.Name = "Mike Sun"
            Dim c3 As New Customer
            c3.ID = 3
            c3.Name = "Jialiang Ge"
            _customercollection.Add(c1)
            _customercollection.Add(c2)
            _customercollection.Add(c3)

        End Sub

        Private Shared _customercollection As ObservableCollection(Of Customer)
        Private Shared _customer As Customer = New Customer()
        Public Shared Function GetSingleCustomer() As Object
            Return _customer
        End Function

        Public Shared Function GetCustomersList() As IEnumerable
            Return _customercollection
        End Function

    End Class

    Public Class Customer
        Implements INotifyPropertyChanged
        Public _id As Integer
        ' Use custom validation class for this property
        <CustomValidation(GetType(MyValidation), "Validate")> _
        Public Property ID() As Integer
            Get
                Return _id
            End Get
            Set(ByVal value As Integer)

                Dim vc As ValidationContext = New ValidationContext(Me, Nothing, Nothing)
                vc.MemberName = "ID"
                Validator.ValidateProperty(value, vc)
                _id = value
                If Not PropertyChangedEvent Is Nothing Then
                    RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs("ID"))
                End If
            End Set
        End Property
        Private _name As String
        ' Use DataAnnotation to restrict the length of Name property
        <StringLength(30, MinimumLength:=1, ErrorMessage:="Name must be between 1 and 30 characters long.")> _
        Public Property Name() As String
            Get
                Return _name
            End Get
            Set(ByVal value As String)

                Dim vc As ValidationContext = New ValidationContext(Me, Nothing, Nothing)
                vc.MemberName = "Name"
                Validator.ValidateProperty(value, vc)
                _name = value
                If Not PropertyChangedEvent Is Nothing Then
                    RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs("Name"))
                End If
            End Set
        End Property

#Region "INotifyPropertyChanged Members"

        Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

#End Region

    End Class

    Public Class MyValidation
        Private Sub New()
        End Sub
        Public Shared Function Validate(ByVal [property] As Object, ByVal context As ValidationContext) As ValidationResult
            ' You can get Property name from context.MemberName. It's not used here for simplicity.
            Dim customerid As Integer
            ' For the simplicity, the change is invalid if new ID equals 11.
            If Int32.TryParse([property].ToString(), customerid) AndAlso customerid = 11 Then
                Dim validationResult As ValidationResult = New ValidationResult("Custom Validation Failed. ID cannot be 11")
                Return validationResult
            Else
                Return ValidationResult.Success
            End If
        End Function
    End Class
End Namespace
